pkg/apis/build/v1beta1/build_conversion.go

```diff
@@ -0,0 +1,381 @@
+	dest.Output.Image = orig.Output.Image
+	if orig.Output.Credentials != nil {
+		dest.Output.PushSecret = &orig.Output.Credentials.Name
+	}
+
+	dest.Output.Annotations = orig.Output.Annotations
+	dest.Output.Labels = orig.Output.Labels
+
+	// Handle BuildSpec Timeout
+	dest.Timeout = orig.Timeout
+
+	// Handle BuildSpec Env
+	dest.Env = orig.Env
+
+	// Handle BuildSpec Retention
+	if orig.Retention != nil {
+		dest.Retention = &BuildRetention{}
+		if orig.Retention != nil {
+			if orig.Retention.FailedLimit != nil {
+				dest.Retention.FailedLimit = orig.Retention.FailedLimit
+			}
+			if orig.Retention.SucceededLimit != nil {
+				dest.Retention.SucceededLimit = orig.Retention.SucceededLimit
+			}
+			if orig.Retention.TTLAfterFailed != nil {
+				dest.Retention.TTLAfterFailed = orig.Retention.TTLAfterFailed
+			}
+			if orig.Retention.TTLAfterSucceeded != nil {
+				dest.Retention.TTLAfterSucceeded = orig.Retention.TTLAfterSucceeded
+			}
+		}
+	}
+
+	// Handle BuildSpec Volumes
+	dest.Volumes = []BuildVolume{}
+	for _, vol := range orig.Volumes {
+		aux := BuildVolume{
+			Name:         vol.Name,
+			VolumeSource: vol.VolumeSource,
+		}
+		dest.Volumes = append(dest.Volumes, aux)
+	}
+
+	return nil
+}
+
+func (dest *BuildSpec) ConvertTo(bs *v1alpha1.BuildSpec) error {
+	// Handle BuildSpec Source
+	bs.Source = getAlphaBuildSource(*dest)
+
+	// Handle BuildSpec Trigger
+	if dest.Trigger != nil {
+		bs.Trigger = &v1alpha1.Trigger{}
+		for _, t := range dest.Trigger.When {
+			tw := v1alpha1.TriggerWhen{}
+			t.convertToAlpha(&tw)
+			bs.Trigger.When = append(bs.Trigger.When, tw)
+		}
+		if dest.Trigger.TriggerSecret != nil {
+			bs.Trigger.SecretRef = &corev1.LocalObjectReference{Name: *dest.Trigger.TriggerSecret}
+		}
+	}
+
+	// Handle BuildSpec Strategy
+	bs.Strategy = v1alpha1.Strategy{
+		Name:       dest.StrategyName(),
+		Kind:       (*v1alpha1.BuildStrategyKind)(dest.Strategy.Kind),
+		APIVersion: dest.Strategy.APIVersion,
+	}
+
+	// Handle BuildSpec Builder, TODO
+	bs.Builder = nil
+
+	// Handle BuildSpec ParamValues
+	bs.ParamValues = nil
+	for _, p := range dest.ParamValues {
+		param := v1alpha1.ParamValue{}
+		p.convertToAlpha(&param)
+		bs.ParamValues = append(bs.ParamValues, param)
+		if param.Name == "PARAM_DOCKERFILE" {
+			bs.Dockerfile = param.Value
+		}
+	}
+
+	// Handle BuildSpec Output
+	// insecure := false
+	// bs.Output.Insecure = &insecure
+	bs.Output.Image = dest.Output.Image
+	if dest.Output.PushSecret != nil {
+		bs.Output.Credentials = &corev1.LocalObjectReference{}
+		bs.Output.Credentials.Name = *dest.Output.PushSecret
+	}
+	bs.Output.Annotations = dest.Output.Annotations
+	bs.Output.Labels = dest.Output.Labels
+
+	// Handle BuildSpec Timeout
+	bs.Timeout = dest.Timeout
+
+	// Handle BuildSpec Env
+	bs.Env = dest.Env
+
+	// Handle BuildSpec Retention
+	if dest.Retention != nil {
+		bs.Retention = &v1alpha1.BuildRetention{}
+		if dest.Retention != nil && dest.Retention.FailedLimit != nil {
```

Comment:
nil checks not needed, you can imo even directly set the fields inside the object constructor.